docs/generic-constraints.md

```diff
@@ -0,0 +1,142 @@
+
+2. Use a `where` clause.
+```luau
+local function getProperty<T, K where K: keyof<T>>( object: T, key: K ): index<T, K>
+	return object[ key ]
+end
+
+type getProperty<T, K where K: keyof<T>> = ( object: T, key: K ) -> ( index<T, K> )
+```
+This would allow users to specify the constraints of the separately from the generic declaration itself. This would be reminiscent to users of C#. Imposing multiple constraints on different generics can be done by delimiting with `,`. This would be backwards compatible, without major performance implications, as it could just be a conditional keyword. 
+
+The `where` clause syntax can be further expanded to pose custom bounds on generics.
+```luau
+local function add<T, K where add<T, K>>( a: T, b: K )
+	return a + b
+end
+```
+The `add` function would only accept `T` and `K` if `add<T, K` is not `never`. This is currently available in Luau, but this behaviour is locked behind `add<T, K>` being the return type.
+
+### Usage
+#### Subtyping Constraints
+This is the problem posed at the beginning of this RFC.
+```lua
+local function callbackProperty<T, K: keyof<T>>( object: T, key: K, callback: (property: index<T, K>) -> () )
+	callback( object[ key ] )
+end
+```
+`K` will be constrained to be `keyof<T>`. This would allow `index<T, K>` to infer the type properly.
+#### Equality Constraints
+Suppose we wanted to make a function that could only add `number`s or `vector`s.
+```luau
+-- Attempt 1
+local function sumSpecific( a: number | vector, b: number | vector ): number | vector
+	return a + b -- not okay!
+end
+-- Attempt 2
+local function sumSpecific<T>( a: T, b: T ): T
+	return a + b -- not okay! not all types are addable.
+end
+```
+This is to be expected. Luau does not interpret `a` and `b` as the same type, but if we use a generic then Luau throws an error that not all types are addable.
+```luau
+-- Attempt 3
+local sumSpecific: ( ( number, number ) -> ( number ) ) & ( ( vector, vector ) -> ( vector ) ) = function( a, b )
+	return a + b -- okay!
+end
+```
+This is fine, but is not concise whatsoever, and does not allow for the traditional function declaration syntax.
+```luau
+local function sumSpecific<T: number | vector>( a: T, b: T ): T
+	return a + b
+end
+
+sumSpecific( vector.create( 42, 42, 42 ) + vector.create( 143, 143, 143 ) ) -- okay!
+sumSpecific( 42, 143 ) -- okay!
+
+sumSpecific( vector.create( 42, 42, 42 ), 143 ) -- error! a and b do not match.
+```
+In this example, `T` must either be a number or vector, and setting annotating variables `a` and `b` as `T` would solve this. This can also be done by overloading the function but is more verbose and less elegant than this solution. This cannot be done with Luau's `add` type function, as it would allow all 'addable' types, and could not guarantee the return type being the same as both inputs.
+
+#### Other Constraints
+```luau
+local function multiply<T, K where T: number, mul<T, K>>( a: T, b: K )
```

Comment:
So if we look closer, we can see that (assuming `return a + b` was meant to be `return a * b`) the function `multiply`'s return type would be inferred as `mul<T, K>`, which is covariant and all its dependent type variables are contravariant: this renders `mul<T, K>` in the `where` clause to be _redundant_. That is, the example is equivalent to

```suggestion
local function multiply<T, K where T: number>( a: T, b: K ): mul<T, K>
```